### PR TITLE
Add support for Tuya air source water heaters (aka heat pump hot water systems)

### PR DESCRIPTION
**Pull request by fsevilla3**
Please accept this pull request that adds support for Tuya air source water heaters (aka heat pump hot water systems) using the water_heater platform. I have tested and validated this with a Hydrotherm DYNAMIC/X8.

### CODE REVIEW

**Comment by fsevilla3** [comment]
Hi @rospogrigio. Checking in to see if you had any feedback on this PR or if there's anything I can do to make your review of it easier? Thank you!

**Comment by fsevilla3** [comment]
I'm working on addressing the findings from `Tox PR CI / 3.8 / ubuntu-latest` and will commit changes shortly.

**Comment by fsevilla3** [comment]
Hi @rospogrigio. I've fixed the python3.8 linting issues but am unable to re-run the linting workflows to validate my fixes. Hopefully should all be okay when you re-run the workflows. Thanks!

**Comment by fsevilla3** [comment]
Thanks @rospogrigio. I've fixed the additional errors found, but can't figure out why `pylint` has issues with the following, which look correct to me. Thanks in advance for your advice.

```
custom_components/localtuya/water_heater.py:6:0: E0611: No name 'WaterHeaterEntityFeature' in module 'homeassistant.components.water_heater' (no-name-in-module)
custom_components/localtuya/water_heater.py:149:20: E1206: Not enough arguments for logging format string (logging-too-few-args)
custom_components/localtuya/water_heater.py:155:12: E1206: Not enough arguments for logging format string (logging-too-few-args)
custom_components/localtuya/water_heater.py:202:16: E1206: Not enough arguments for logging format string (logging-too-few-args)
```

File: custom_components/localtuya/water_heater.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+from homeassistant.components.water_heater import (
+    DOMAIN,
+    STATE_ECO,
+    STATE_ELECTRIC,
+    STATE_GAS,
+    STATE_HEAT_PUMP,
+    STATE_HIGH_DEMAND,
+    STATE_PERFORMANCE,
+    WaterHeaterEntity,
+    WaterHeaterEntityFeature
```
Comment (rospogrigio) [inline custom_components/localtuya/water_heater.py:15]: don't know why it does not find the WaterHeaterEntityFeature name... maybe different HA version?

**Review by rospogrigio — CHANGES_REQUESTED** [review]

File: custom_components/localtuya/water_heater.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+                self._attr_current_operation = STATE_OFF
+            elif self.dps_conf(CONF_MODE_DP) is None:
+                _LOGGER.error(
+                    "CONF_MODE_DP has no value. Defaulting to STATE_OFF."
+                )
+                self._attr_current_operation = STATE_OFF
+            else:
+                _LOGGER.debug(
+                    "Device status updated. Tuya: [%s]. HA: [%s]", (
+                        self.dps_conf(CONF_MODE_DP),
```
[the remark this answers is not in the document]
Comment (fsevilla3) [reply custom_components/localtuya/water_heater.py]: Fixed in [31852d8](https://github.com/rospogrigio/localtuya/pull/982/commits/31852d899b5fb7f47844f91095b6ee3e9d1cf452).

File: custom_components/localtuya/water_heater.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+                if operation_mode != self._attr_current_operation:
+                    _LOGGER.debug(
+                        "Changing operation mode. Tuya: [%s]. HA: [%s]",
+                        (op_mode_tuya, operation_mode)
+                    )
+                    await self._device.set_dp(op_mode_tuya, self._conf_mode_dp)
+        else:
+            _LOGGER.error(
+                "Invalid operation mode. Tuya: [%s]. HA: [%s].",
+                (op_mode_tuya, operation_mode)
```
[the remark this answers is not in the document]
Comment (fsevilla3) [reply custom_components/localtuya/water_heater.py]: Fixed in [31852d8](https://github.com/rospogrigio/localtuya/pull/982/commits/31852d899b5fb7f47844f91095b6ee3e9d1cf452).

File: custom_components/localtuya/water_heater.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+                        "Ignoring request to switch off device that's already off."
+                    )
+            else:
+                if self._attr_current_operation == STATE_OFF:
+                    _LOGGER.debug("Switching on device.")
+                    await self._device.set_dp(True, self._dp_id)
+                if operation_mode != self._attr_current_operation:
+                    _LOGGER.debug(
+                        "Changing operation mode. Tuya: [%s]. HA: [%s]",
+                        (op_mode_tuya, operation_mode)
```
[the remark this answers is not in the document]
Comment (fsevilla3) [reply custom_components/localtuya/water_heater.py]: Fixed in [31852d8](https://github.com/rospogrigio/localtuya/pull/982/commits/31852d899b5fb7f47844f91095b6ee3e9d1cf452).

File: custom_components/localtuya/water_heater.py (hunk 1 above)
Comment (fsevilla3) [reply custom_components/localtuya/water_heater.py:15]: Yes, that might be the case. Would you know which version the workflow checks compare against? I'm using HA 2022.8.5 and this integration works fine with it. I've also checked the HA GitHub and `WaterHeaterEntityFeature` is in both the [2022.8.5 release](https://github.com/home-assistant/core/blob/2022.8.5/homeassistant/components/water_heater/__init__.py#L60) and [dev (default) branch](https://github.com/home-assistant/core/blob/dev/homeassistant/components/water_heater/__init__.py#L60).

**Comment by fsevilla3** [comment]
Hi @rospogrigio. Lots of errors with the CI/CD checks when `requirements_test.txt` was asking for the latest HA, so I looked through the history and [home-assistant 2022.5.0](https://github.com/home-assistant/core/blob/2022.5.0/homeassistant/components/water_heater/__init__.py#L60) appears to be the oldest version that has the `WaterHeaterEntityFeature` enum available. Are you happy to approve the re-running of the checks to see if this resolves things with this PR? Thank you!

**Comment by rospogrigio** [comment]
I cant't run the checks, you must resolve the conflict that has raised first...

**Comment by fsevilla3** [comment]
@rospogrigio Whoops, sorry, done. CI/CD workflows awaiting approval now. Thanks!

**Pull request closed without merging**